### ISSUE TITLE
libmali

### ISSUE THREAD

**Issue by asdf288** [libretro/mupen64plus-libretro-nx#288]

**Comment by asdf288** [comment]
Here is the error message generated by Retroarch in debug mode:

```
[INFO] [Playlist]: Loading favorites file: [/storage/.config/retroarch/content_favorites.lpl].
[INFO] [GL]: VSync => on
[INFO] [Environ]: SYSTEM_DIRECTORY: "/tmp/system".
[libretro INFO] mupen64plus: Game controller 0 (Standard controller) has a Memory pak plugged in
[libretro INFO] mupen64plus: Game controller 1 (Standard controller) has nothing plugged in
[libretro INFO] mupen64plus: Game controller 2 (Standard controller) has nothing plugged in
[libretro INFO] mupen64plus: Game controller 3 (Standard controller) has nothing plugged in
[libretro INFO] mupen64plus: Using CIC type X102
[INFO] [Environ]: SYSTEM_DIRECTORY: "/tmp/system".
[INFO] [Environ]: SYSTEM_DIRECTORY: "/tmp/system".
[INFO] [Environ]: SYSTEM_DIRECTORY: "/tmp/system".
[INFO] [Environ]: SYSTEM_DIRECTORY: "/tmp/system".
AddressSanitizer:DEADLYSIGNAL
=================================================================
==908==ERROR: AddressSanitizer: SEGV on unknown address 0x00000000 (pc 0x00000000 bp 0x00000000 sp 0xc72915a0 T0)
==908==Hint: pc points to the zero page.
==908==The signal is caused by a READ memory access.
==908==Hint: address points to the zero page.
    #0 0x0  (<unknown module>)

AddressSanitizer can not provide additional info.
SUMMARY: AddressSanitizer: SEGV (<unknown module>)
==908==ABORTING
```

**Comment by m4xw** [comment]
Can you check if it takes the early return or if the frontend cb returns null?
![image](https://user-images.githubusercontent.com/13141469/102782319-72cc9700-4399-11eb-86d4-9a45329bd9db.png)

This is most likely a frontend ctx driver bug

**Comment by m4xw** [comment]
If the actual function call returns NULL, then i need to know which symbol.
I will need a trace ideally if its a NULLptr GL func thats called so I can identify it.

**Comment by asdf288** [comment]
Thank you for the instructions. I have never debugged a linux program before, are these steps that I followed correct?

- gdb /bin/retroarch
- break glsm_get_proc_address
- run

The problem I'm having is that the system freezes as soon as I start the core in Retroarch, so the breakpoint is never reached. I tried without the breakpoint, but it's the same problem. I will try some more and post my results here.

**Comment by m4xw** [comment]
just:

gdb retroarch
r --verbose
<run as normal, wait till crash>
info file
info regs
bt full

**Comment by asdf288** [comment]
Sorry, it took a long time because I had to recompile Retroarch, the debugging symbols were missing. Here is the output of the gdb commands:

https://pastebin.com/7ASeAUre

By stepping through the code, I could see that the `glsm_get_proc_address` function went through all the symbols without errors. It then went to retroarch.c and there, in line 34612, the crash occurred. At least that's when gdb stopped and I could not issue any further step commands.

**Comment by m4xw** [comment]
The glGetString ptr is null, thats bad.
Try adding -DGL_USE_DLSYM to COREFLAGS for the platform and make clean

**Comment by asdf288** [comment]
This lead to an error that RTLD_DEFAULT could not be found, I searched for this variable and found it defined in dlfcn.h. I added the include directive and recompiled, but the error is exactly the same as before, the backtrace is identical as far as I can see.

**Comment by asdf288** [comment]
Out of curiosity, where in the gdb output did you see that the pointer to glGetString is null? Is it just that symbol or are all pointers null?

I have tried the following: `strings /usr/lib/libmali.so | grep glGetString`
This resulted in two matches, so at least this symbol exists somewere in the library

**Comment by m4xw** [comment]
> #1  0xe4701450 in opengl::FunctionWrapper::wrGetString 
Without thr it will call ptrGetString, which is NULL for you

**Comment by asdf288** [comment]
I suspected that the driver returns a zero value for GL_VERSION. I edited the source code here:
https://git.libretro.com/libretro/mupen64plus-libretro-nx/-/blob/6ee49fd23804e15d7fa6d01e6e5b7f15da760056/GLideN64/src/Graphics/OpenGLContext/ThreadedOpenGl/opengl_Wrapper.cpp#L324
To make it return a fixed string "OpenGL ES 2.0". This seemed to work and the other GL strings were there, but then another error appeared in wrGetIntegerv, this time it was GL_MAJOR_VERSION that was missing. I did not follow this any further until I test another version of the libmali driver to see if that fixes it. Maybe something is wrong with this version and it does not correctly report its GL features when queried.

**Comment by asdf288** [comment]
Update: Recompiling the core with "-DGL_USE_DLSYM" did indeed work! Thank you. It just did not compile correctly on my first try. So to fix this issue, here are the steps needed:

Add the following include directive to [GLFunctions.cpp](https://github.com/libretro/mupen64plus-libretro-nx/blob/develop/GLideN64/src/Graphics/OpenGLContext/GLFunctions.cpp):
`#include <dlfcn.h>`

Compile the core with the following option (add your other platform flags accordingly):
`make CPUFLAGS="-DGL_USE_DLSYM"`

Note: Setting COREFLAGS in this fashion did not work for some reason, it makes the compilation fail.

**Comment by m4xw** [comment]
I wonder why raspberry doesnt need the dlfcn include there, probably in some header hmmmmmm.
Regarding COREFLAGS, you need to set it in your platform target in the makefile, where did u make the change before?
While it can work passing it via the command, i highly prefer all those things explicitly stated in a segment of the makefile

**Comment by asdf288** [comment]
Yes, I did not try to add it in the makefile directly yet. I'll do that now. So far, I tried `make COREFLAGS="-DGL_USE_DLSYM"` but this seems to break something in the Makefile, a lot of the usual flags are suddenly missing and compilation fails.

Also, I wasn't sure if the Makefile is the best place to put it, because this problem seems to be driver-specific, not platform specific, and the Makefile does not know which driver is actually being used. Many platforms have at least two possible OpenGL drivers (Mesa and another platform-specific driver).

**Comment by m4xw** [comment]
> I tried make COREFLAGS="-DGL_USE_DLSYM" but this seems to break something in the Makefile, a lot of the usual flags are suddenly missing and compilation fails.

I never tried that.. so not supported :P

> Also, I wasn't sure if the Makefile is the best place to put it, because this problem seems to be driver-specific, not platform specific, and the Makefile does not know which driver is actually being used. Many platforms have at least two possible OpenGL drivers (Mesa and another platform-specific driver).

Fwiw this could be a frontend issue in the context like I said, it's kinda what the hack works-around.
It should work as long any shared library that exposes the gl funcs is loaded.
I used to call it YOLO_DLSYM :P
